Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CaloRawFitter.h

```diff
@@ -0,0 +1,156 @@
+  /// \return index of max aplitude in array,
+  /// \return pedestal,
+  /// \return first time bin,
+  /// \return last time bin,
+  /// \return ampitude cut
+  std::tuple<int, int, float, short, short, float, int, int> preFitEvaluateSamples(const std::vector<AltroDecoder::Bunch>& bunchvector,
+                                                                                   std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2, int acut);
+
+  /// \brief The require time range if the maximum ADC value is between min and max (timebin)
+  void setTimeConstraint(int min, int max);
+
+  void setIsZeroSuppressed(bool iszs = true) { mIsZerosupressed = iszs; }
+  void setAmpCut(float cut) { mAmpCut = cut; }
+  void setNsamplePed(int i) { mNsamplePed = i; }
+  void setL1Phase(double phase) { mL1Phase = phase; }
+
+  bool getIsZeroSuppressed() const { return mIsZerosupressed; }
+  float getAmpCut() const { return mAmpCut; }
+  int getNsamplePed() const { return mNsamplePed; }
+
+  // access to array info
+  double getReversed(const int i) const { return mReversed[i]; }
+  const char* getAlgoName() const { return mName.c_str(); }
+  const char* getAlgoAbbr() const { return mNameShort.c_str(); }
+  FitAlgorithm getAlgo() const { return mAlgo; }
+
+  /// \brief Get the maximum of a bunch array
+  /// \return Maximum amplitute
+  short maxAmp(const AltroDecoder::Bunch* bunch, int* maxindex) const;
+
+  /// \brief Get maximum of array
+  /// \return Maximum amplitute
+  unsigned short maxAmp(const unsigned short* data, const int length) const;
+
+  /// \brief A bunch is considered invalid if the maximum is in the first or last time-bin.
+  bool checkBunchEdgesForMax(const AltroDecoder::Bunch* const bunch) const;
+
+  /// \brief Check if the index of the max ADC vaue is consistent with trigger.
+  bool isInTimeRange(const int maxindex, const int maxtime, const int mintime) const;
+
+  /// \brief Time sample comes in reversed order, revers them back Subtract the baseline based on content of altrocfg1 and altrocfg2.
+  /// \return Pedestal
+  float reverseAndSubtractPed(const AltroDecoder::Bunch* bunch,
+                              std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2,
+                              double* outarray) const;
+  /// \brief We select the bunch with the highest amplitude unless any time constraints is set.
+  /// \return The index of the array with the maximum aplitude
+  int selectBunch(const std::vector<AltroDecoder::Bunch>& bunchvector,
+                  short* maxampbin, short* maxamplitude);
+
+  /// \brief Selection of subset of data from one bunch that will be used for fitting or Peak finding.
+  /// Go to the left and right of index of the maximum time bin
+  /// Until the ADC value is less than cut, or derivative changes sign (data jump)
+  void selectSubarray(const double* data, int length, short maxindex,
+                      int* first, int* last, int cut) const;
```

Comment:
why not using ``int& first, int& last``?

Reply:
At least the first data field (data + length) should be replace by gsl::span, probably the two pointers as well.